Preact/compat resolve alias issue with external React component dependency when using SSR

### Describe the bug

I've been trying to get a Vite SSR setup with vite-plugin-ssr going recently using Preact, and ran into an issue with a `react -> preact/compat` resolve failing when trying to use a React component library ([react-icons](https://react-icons.github.io/react-icons/) in this case).

Some notes on the reproduction repo contents:

- The setup is based on the react-ts vite-plugin-ssr init.
- `@vite/plugin-react` has been swapped to [`@preact/preset-vite`](https://github.com/preactjs/preset-vite) in vite.config.ts. This preact preset sets up `resolve.alias` for `react -> preact/compat` and `react-dom -> preact/compat`.
- Running vite's own dev server (either via `vite` on CLI or via the JS API as seen in [`server-vite.js`](https://github.com/Daiz/vps-preact-compat-bug/blob/master/server-vite.js)) works just fine.
- Running vite in `middlewareMode: "ssr"` and with express works just fine too. (see [`server-express.js`](https://github.com/Daiz/vps-preact-compat-bug/blob/master/server-express.js)).
- It's just when using the vite-plugin-ssr server setup ([`server-vps.js`](https://github.com/Daiz/vps-preact-compat-bug/blob/master/server-vps.js)) where trying to get the index page results in a resolve error.
- One thing unique to the vite-plugin-ssr setup compared to the others is the usage of [preact-render-to-string](https://github.com/preactjs/preact-render-to-string) in [`renderer/_default.page.server.tsx`](https://github.com/Daiz/vps-preact-compat-bug/blob/master/renderer/_default.page.server.tsx) but I also tried changing [`src/main.tsx`](https://github.com/Daiz/vps-preact-compat-bug/blob/master/src/main.tsx) to use this module for rendering (and setting with innerHTML) in the vite/express dev server setup and that works just fine.
- I found the [SSR Externals](https://vitejs.dev/guide/ssr.html#ssr-externals) page in Vite docs and tried various `ssr: { external: [], noExternal: [] }` configurations with `"react-icons"` and `"react-icons/fa"`, but this didn't seem to have any effect on the problem.

By all indications I don't think there should be anything wrong with the setup in itself, so it feels like *something* is going awry with Vite SSR specifically here to cause the issue. Perhaps the [SSR module loader](https://github.com/vitejs/vite/blob/4a764f52e4964b02c02f1ce6863ae3454daad55c/packages/vite/src/node/ssr/ssrModuleLoader.ts#L209-L281) isn't respecting resolve aliases properly?

### Reproduction

https://github.com/Daiz/vps-preact-compat-bug

### System Info

```shell
System:
    OS: Windows 10 10.0.19043
    CPU: (16) x64 AMD Ryzen 7 3800X 8-Core Processor
    Memory: 8.54 GB / 31.93 GB
  Binaries:
    Node: 14.17.5 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.11 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 96.0.4664.110
    Edge: Spartan (44.19041.1266.0), Chromium (97.0.1072.62)
    Internet Explorer: 11.0.19041.1202
  npmPackages:
    vite: ^2.7.12 => 2.7.12
```


### Used Package Manager

npm

### Logs

```shell
Server running at http://localhost:1337
[vite] Error when evaluating SSR module /pages/index.page.tsx:
Error: Cannot find module 'react' imported from 'I:\Daiz\vps-preact-compat-bug\node_modules\react-icons\lib\cjs\iconBase.js'
    at viteResolve (I:\Daiz\vps-preact-compat-bug\node_modules\vite\dist\node\chunks\dep-0351185a.js:60123:25)
    at Function.<anonymous> (I:\Daiz\vps-preact-compat-bug\node_modules\vite\dist\node\chunks\dep-0351185a.js:60137:20)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (I:\Daiz\vps-preact-compat-bug\node_modules\react-icons\lib\cjs\iconBase.js:26:13)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
```


### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/docs/config/shared-options.md
+++ b/docs/config/shared-options.md
@@ -110,6 +110,10 @@ When aliasing to file system paths, always use absolute paths. Relative alias va
 
 More advanced custom resolution can be achieved through [plugins](/guide/api-plugin).
 
+::: warning Using with SSR
+If you have configured aliases for [SSR externalized dependencies](/guide/ssr.md#ssr-externals), you may want to alias the actual `node_modules` packages. Both [Yarn](https://classic.yarnpkg.com/en/docs/cli/add/#toc-yarn-add-alias) and [pnpm](https://pnpm.js.org/en/aliases) support aliasing via the `npm:` prefix.
+:::
+
 ## resolve.dedupe
 
 - **Type:** `string[]`

